src/render/python/shape_v.cpp

```diff
@@ -217,6 +219,75 @@ template <typename Ptr, typename Cls> void bind_shape_generic(Cls &cls) {
+    MI_PY_IMPORT_TYPES()
+    cls
+       .def("vertex_count", [](const Ptr ptr) {
+            return ptr->vertex_count();
+        }, D(Mesh, vertex_count))
+       .def("face_count", [](const Ptr ptr) {
+            return ptr->face_count();
+        }, D(Mesh, face_count))
+       .def("has_vertex_normals", [](const Ptr ptr) {
+            return ptr->has_vertex_normals();
+        }, D(Mesh, has_vertex_normals))
+       .def("has_vertex_texcoords", [](const Ptr ptr) {
+            return ptr->has_vertex_texcoords();
+        }, D(Mesh, has_vertex_texcoords))
+       .def("has_mesh_attributes", [](const Ptr ptr) {
+            return ptr->has_mesh_attributes();
+        }, D(Mesh, has_mesh_attributes))
+       .def("has_face_normals", [](const Ptr ptr) {
+            return ptr->has_face_normals();
+        }, D(Mesh, has_face_normals))
+       .def("face_indices", [](const Ptr m, UInt32 index, Mask active) {
+                return m->face_indices(index, active);
+            }, D(Mesh, face_indices), "index"_a, "active"_a = true)
+       .def("edge_indices", [](const Ptr m, UInt32 tri_index, UInt32 edge_index, Mask active) {
+                return m->edge_indices(tri_index, edge_index, active);
+            }, D(Mesh, edge_indices), "tri_index"_a, "edge_index"_a, "active"_a = true)
+       .def("vertex_position", [](const Ptr m, UInt32 index, Mask active) {
+                return m->vertex_position(index, active);
+            }, D(Mesh, vertex_position), "index"_a, "active"_a = true)
+       .def("vertex_normal", [](const Ptr m, UInt32 index, Mask active) {
+                return m->vertex_normal(index, active);
+            }, D(Mesh, vertex_normal), "index"_a, "active"_a = true)
+       .def("vertex_texcoord", [](const Ptr m, UInt32 index, Mask active) {
+                return m->vertex_texcoord(index, active);
+            }, D(Mesh, vertex_texcoord), "index"_a, "active"_a = true)
+       .def("face_normal", [](const Ptr m, UInt32 index, Mask active) {
+                return m->face_normal(index, active);
+            }, D(Mesh, face_normal), "index"_a, "active"_a = true)
+       .def("opposite_dedge", [](const Ptr m, UInt32 index, Mask active) {
+                return m->opposite_dedge(index, active);
+            }, D(Mesh, opposite_dedge), "index"_a, "active"_a = true)
+
+       .def("ray_intersect_triangle", [](const Ptr ptr, const UInt32 &index,
+                                         const Ray3f &ray, Mask active) {
+                ptr->ray_intersect_triangle(index, ray, active);
+            },
+            "index"_a, "ray"_a, "active"_a = true,
+            D(Mesh, ray_intersect_triangle));
+
+    if constexpr (dr::is_array_v<Ptr> && dr::is_jit_v<Ptr>) {
+        // Custom constructor to automatically zero-out non-Mesh pointer entries.
+        // using ShapePtr = dr::replace_scalar_t<Ptr, Shape *>;
+        cls
+            .def("__init__", [](Ptr *dst, const ShapePtr &ptr) {
```

Comment:
Ideally I would have wanted to add extra constructors for `const ShapePtr &` and leave the basic ones as-is, but I don't think it's possible?

Reply:
I'm not sure I understand. Are you talking about `.def("__init__", [](Ptr *dst, const Mesh *ptr)` ? As in, it should already exist?

Although unfortunate, this is expected because of how the original Dr.Jit contstructors are defined (they are defined with [type slots](https://nanobind.readthedocs.io/en/latest/typeslots.html)). They will be lost when using the `def("__init__", ... )` mechanism.

Reply:
Okay, makes sense! It's not too much code to add back the constructors we need.